### PR TITLE
check for c++17 features used in shogun

### PR DESCRIPTION
**Pull request by gf712**
added checks for c++17 features:
 - if constexpr
 - folding expressions
 - init-statements for if

let me know what else you need

### CODE REVIEW

File: src/shogun/CMakeLists.txt [hunk 1]
```diff
@@ -351,6 +351,50 @@ IF (NOT HAVE_STD_VARIANT)
...
+auto func(Args... args) {
+    return (args + ...);
+}
+int main()
+{return func(1,2,3);}
+" HAVE_FOLDING_EXPRESSIONS)
+
+IF(NOT HAVE_FOLDING_EXPRESSIONS)
+  MESSAGE(FATAL_ERROR "Compiling shogun requires a compiler with folding expression support!")
+ENDIF()
```
Comment (gf712) [inline src/shogun/CMakeLists.txt]: right now this doesn't have to be fatal, but would be good to be able to rely on this

File: src/shogun/CMakeLists.txt [hunk 2]
```diff
@@ -351,6 +351,50 @@ IF (NOT HAVE_STD_VARIANT)
...
+    return value;
+  }
+  else
+  {
+    return value - 1;
+  }
+}
+" HAVE_IF_INIT)
+IF(NOT HAVE_IF_INIT)
+  MESSAGE(FATAL_ERROR "Compiling shogun requires a compiler with if init-statements support!")
```
Comment (gf712) [inline src/shogun/CMakeLists.txt]: same here

**Comment by lisitsyn** [comment]
Yes!

**Comment by gf712** [comment]
Yes, need clang 4 and gcc 7 in the dockerfile. But there is no stable support for gcc 7 in Debian yet..

**Comment by lisitsyn** [comment]
> Yes, need clang 4 and gcc 7 in the dockerfile. But there is no stable support for gcc 7 in Debian yet..

Why Debian? I think we can install them from PPA in Ubuntu 16

**Comment by gf712** [comment]
> Why Debian? I think we can install them from PPA in Ubuntu 16
I think the dockerfile uses debian. And the CI in linux uses the docker image no? need to check..

File: .ci/ci.yml [hunk 3]
```diff
@@ -39,7 +39,7 @@ jobs:
       clang:
         CC: "clang-4.0"
         CXX: "clang++-4.0"
-        cmakeOptions: ''
+        cmakeOptions: '-DCLANG_DEFAULT_CXX_STDLIB=libc++'
```
Comment (vigsterkr) [inline .ci/ci.yml]: this is just a hack... just saying... plz dont ever merge it.

Comment (gf712) [reply .ci/ci.yml]: yup just checking if this is the issue

**Comment by vigsterkr** [comment]
> > Yes, need clang 4 and gcc 7 in the dockerfile. But there is no stable support for gcc 7 in Debian yet..
> 
> Why Debian? I think we can install them from PPA in Ubuntu 16

if you upgrade donbot to ubuntu 16 etc then go ahead... see our buildbot infra

**Comment by lisitsyn** [comment]
Oh I see what you mean :)

**Comment by vigsterkr** [comment]
FYI
```/usr/include/python2.7/unicodeobject.h:534:5: error: ISO C++17 does not allow 'register' storage class specifier [-Wregister]
    register PyObject *obj,     /* Object */
    ^~~~~~~~~
/usr/include/python2.7/unicodeobject.h:553:5: error: ISO C++17 does not allow 'register' storage class specifier [-Wregister]
    register PyObject *obj      /* Object */
    ^~~~~~~~~
/usr/include/python2.7/unicodeobject.h:575:5: error: ISO C++17 does not allow 'register' storage class specifier [-Wregister]
    register const wchar_t *w,  /* wchar_t buffer */
    ^~~~~~~~~
/usr/include/python2.7/unicodeobject.h:593:5: error: ISO C++17 does not allow 'register' storage class specifier [-Wregister]
    register wchar_t *w,        /* wchar_t buffer */
    ^~~~~~~~~
In file included from /home/wiking/shogun/build/src/interfaces/python/shogunPYTHON_wrap.cxx:175:
In file included from /usr/include/python2.7/Python.h:94:
/usr/include/python2.7/stringobject.h:173:5: error: ISO C++17 does not allow 'register' storage class specifier [-Wregister]
    register PyObject *obj,     /* string or Unicode object */
    ^~~~~~~~~
/usr/include/python2.7/stringobject.h:174:5: error: ISO C++17 does not allow 'register' storage class specifier [-Wregister]
    register char **s,          /* pointer to buffer variable */
    ^~~~~~~~~
/usr/include/python2.7/stringobject.h:175:5: error: ISO C++17 does not allow 'register' storage class specifier [-Wregister]
    register Py_ssize_t *len    /* pointer to length variable or NULL
    ^~~~~~~~~```

so basically python2.7 is bye bye bye....

**Comment by karlnapf** [comment]
Kill it, think official support stops soon as well

**Pull request merged**